File: src/movieclaw_api/schemas/channels.py
"""IM 通道(微信绑定)相关的请求/响应模型。"""

from __future__ import annotations

from datetime import datetime

from pydantic import Field

from movieclaw_api.schemas.base import BaseModel
from movieclaw_db.models.channel_account import ChannelAccount


class WeixinAccountView(BaseModel):
    """已绑定账号(绑定页列表项)。"""

    account_id: str
    #: 扫码绑定人的微信用户 id(白名单)
    bound_user_id: str | None
    #: active=正常;stale=凭据失效,需重新扫码
    status: str
    #: 当前进程内收发循环是否在运行
    running: bool
    last_error: str | None
    bound_at: datetime

    @classmethod
    def from_model(cls, row: ChannelAccount, *, running: bool) -> WeixinAccountView:
        return cls(
            account_id=row.account_id,
            bound_user_id=row.bound_user_id,
            status=row.status,
            running=running,
            last_error=row.last_error,
            bound_at=row.created_at,
        )


class WeixinBindingStartView(BaseModel):
    """发起绑定的返回:前端渲染二维码并开始 poll。"""

    challenge_id: str
    #: 二维码内容链接(扫码目标;备用展示)
    qrcode_url: str
    #: 服务端渲染好的二维码图(SVG data URL),前端 <img> 直接显示
    qrcode_image: str
    message: str


class WeixinBindingStatusView(BaseModel):
    """绑定状态快照(前端每 1-2 秒 poll 一次,后端只读内存,毫秒级返回)。

    status 取值:
    pending / scanned / need_verify_code / confirmed / already_bound /
    expired / failed。confirmed 时通道已启动,account 字段附带账号信息。
    """

    challenge_id: str
    status: str
    message: str
    #: 二维码可能中途刷新(过期自动换新),前端每次 poll 都以此为准重绘
    qrcode_url: str
    qrcode_image: str
    account: WeixinAccountView | None = None


class WeixinVerifyCodePayload(BaseModel):
    """提交手机微信上显示的配对数字。"""

    code: str = Field(min_length=1, max_length=16, description="配对码")


# ---------------------------------------------------------------------------
# Telegram / Discord(配对码绑定)
# ---------------------------------------------------------------------------


class ImAccountView(BaseModel):
    """已绑定的 TG/Discord bot 账号(绑定页列表项)。"""

    channel_id: str
    account_id: str
    #: 完成配对的用户 id(白名单,同时是推送目标)
    bound_user_id: str | None
    status: str
    running: bool
    last_error: str | None
    bound_at: datetime

    @classmethod
    def from_model(cls, row: ChannelAccount, *, running: bool) -> ImAccountView:
        return cls(
            channel_id=row.channel_id,
            account_id=row.account_id,
            bound_user_id=row.bound_user_id,
            status=row.status,
            running=running,
            last_error=row.last_error,
            bound_at=row.created_at,
        )


class ImBindTokenPayload(BaseModel):
    """发起绑定:提交 bot token。"""

    token: str = Field(min_length=8, max_length=256, description="bot token")


class FeishuBindPayload(BaseModel):
    """接入飞书群自定义机器人(粘贴 Webhook 地址即绑即用,无配对码)。"""

    webhook_url: str = Field(
        min_length=16, max_length=512, description="飞书自定义机器人 Webhook 地址"
    )
    secret: str = Field(default="", max_length=256, description="签名校验密钥;未开启签名校验留空")


class ImBindingView(BaseModel):
    """配对绑定状态(发起返回 + 前端 poll 同一结构)。

    status 取值:pending / confirmed / expired / failed。
    """

    challenge_id: str
    status: str
    #: 面板展示给用户的 6 位配对码(用户私聊 bot 发这串数字完成绑定)
    pair_code: str
    bot_name: str
    message: str
    account: ImAccountView | None = None


class PushTestPayload(BaseModel):
    """测试推送文本(缺省用默认文案)。"""

    text: str = Field(default="", max_length=500)


class ChannelPushConfigView(BaseModel):
    """推送内容开关(GET 返回与 PUT 载荷同构)。"""

    push_dispatch: bool = Field(description="订阅开始下载时推送")
    push_imported: bool = Field(description="入库完成时推送")

File: apps/web/lib/api/channels.ts
import { request } from "@/lib/http";

/** 后端统一响应信封（见 movieclaw_api.schemas.response.ApiResponse） */
interface ApiEnvelope<T> {
  success: boolean;
  code: string;
  message: string;
  data: T;
}

async function unwrap<T>(promise: Promise<ApiEnvelope<T>>): Promise<T> {
  return (await promise).data;
}

/** 已绑定的微信账号（见 schemas.channels.WeixinAccountView）。 */
export interface WeixinAccount {
  account_id: string;
  /** 扫码绑定人的微信用户 id（白名单：只有此人能对话） */
  bound_user_id: string | null;
  /** active=正常；stale=凭据失效需重新扫码 */
  status: "active" | "stale";
  /** 当前进程内收发循环是否在运行 */
  running: boolean;
  last_error: string | null;
  bound_at: string;
}

/** 绑定流程状态（见 schemas.channels.WeixinBindingStatusView 的注释）。 */
export type WeixinBindingStatus =
  | "pending"
  | "scanned"
  | "need_verify_code"
  | "confirmed"
  | "already_bound"
  | "expired"
  | "failed";

export interface WeixinBindingStart {
  challenge_id: string;
  qrcode_url: string;
  /** 服务端渲染好的二维码 SVG data URL，<img> 直接显示 */
  qrcode_image: string;
  message: string;
}

export interface WeixinBindingSnapshot {
  challenge_id: string;
  status: WeixinBindingStatus;
  message: string;
  qrcode_url: string;
  qrcode_image: string;
  account: WeixinAccount | null;
}

export function listWeixinAccounts(init?: RequestInit): Promise<WeixinAccount[]> {
  return unwrap(request<ApiEnvelope<WeixinAccount[]>>("/channels/weixin/accounts", init));
}

export function startWeixinBinding(): Promise<WeixinBindingStart> {
  return unwrap(
    request<ApiEnvelope<WeixinBindingStart>>("/channels/weixin/bindings", { method: "POST" }),
  );
}

export function getWeixinBindingStatus(challengeId: string): Promise<WeixinBindingSnapshot> {
  return unwrap(
    request<ApiEnvelope<WeixinBindingSnapshot>>(
      `/channels/weixin/bindings/${encodeURIComponent(challengeId)}`,
    ),
  );
}

export function submitWeixinVerifyCode(
  challengeId: string,
  code: string,
): Promise<Record<string, never>> {
  return unwrap(
    request<ApiEnvelope<Record<string, never>>>(
      `/channels/weixin/bindings/${encodeURIComponent(challengeId)}/verify-code`,
      { method: "POST", body: JSON.stringify({ code }) },
    ),
  );
}

export function unbindWeixinAccount(accountId: string): Promise<Record<string, never>> {
  return unwrap(
    request<ApiEnvelope<Record<string, never>>>(
      `/channels/weixin/accounts/${encodeURIComponent(accountId)}`,
      { method: "DELETE" },
    ),
  );
}

// ---------------------------------------------------------------------------
// Telegram / Discord / 飞书（见 api/routes/channels_im.py）
// ---------------------------------------------------------------------------

export type ImChannelId = "telegram" | "discord" | "feishu";

/** 已绑定的 TG/Discord/飞书账号（见 schemas.channels.ImAccountView）。 */
export interface ImAccount {
  channel_id: string;
  account_id: string;
  /** 完成配对的用户 id（白名单，同时是推送目标） */
  bound_user_id: string | null;
  status: "active" | "stale";
  running: boolean;
  last_error: string | null;
  bound_at: string;
}

/** 配对绑定状态（发起返回与轮询同一结构）。 */
export interface ImBinding {
  challenge_id: string;
  status: "pending" | "confirmed" | "expired" | "failed";
  /** 面板展示的 6 位配对码：用户私聊 bot 发这串数字完成绑定 */
  pair_code: string;
  bot_name: string;
  message: string;
  account: ImAccount | null;
}

export function listImAccounts(channel: ImChannelId): Promise<ImAccount[]> {
  return unwrap(request<ApiEnvelope<ImAccount[]>>(`/channels/im/${channel}/accounts`));
}

export function startImBinding(channel: ImChannelId, token: string): Promise<ImBinding> {
  return unwrap(
    request<ApiEnvelope<ImBinding>>(`/channels/im/${channel}/bindings`, {
      method: "POST",
      body: JSON.stringify({ token }),
    }),
  );
}

export function getImBindingStatus(channel: ImChannelId, challengeId: string): Promise<ImBinding> {
  return unwrap(
    request<ApiEnvelope<ImBinding>>(
      `/channels/im/${channel}/bindings/${encodeURIComponent(challengeId)}`,
    ),
  );
}

export function unbindImAccount(
  channel: ImChannelId,
  accountId: string,
): Promise<Record<string, never>> {
  return unwrap(
    request<ApiEnvelope<Record<string, never>>>(
      `/channels/im/${channel}/accounts/${encodeURIComponent(accountId)}`,
      { method: "DELETE" },
    ),
  );
}

/**
 * 接入飞书群自定义机器人：粘贴 Webhook 地址即绑即用（服务端发欢迎消息验真，
 * 无配对码、无轮询）。secret 为签名校验密钥，未开启签名校验传空。
 */
export function startFeishuBinding(webhookUrl: string, secret: string): Promise<ImAccount> {
  return unwrap(
    request<ApiEnvelope<ImAccount>>(`/channels/im/feishu/bindings`, {
      method: "POST",
      body: JSON.stringify({ webhook_url: webhookUrl, secret }),
    }),
  );
}

export function sendChannelPushTest(text?: string): Promise<{ sent: number }> {
  return unwrap(
    request<ApiEnvelope<{ sent: number }>>(`/channels/im/push-test`, {
      method: "POST",
      body: JSON.stringify({ text: text ?? "" }),
    }),
  );
}

/** 推送内容开关（哪些系统事件会推送到已绑定通道）。 */
export interface ChannelPushConfig {
  /** 订阅命中并投递下载时推送 */
  push_dispatch: boolean;
  /** 下载完成整理入库时推送 */
  push_imported: boolean;
}

export function getChannelPushConfig(): Promise<ChannelPushConfig> {
  return unwrap(request<ApiEnvelope<ChannelPushConfig>>(`/channels/im/push-config`));
}

export function updateChannelPushConfig(config: ChannelPushConfig): Promise<ChannelPushConfig> {
  return unwrap(
    request<ApiEnvelope<ChannelPushConfig>>(`/channels/im/push-config`, {
      method: "PUT",
      body: JSON.stringify(config),
    }),
  );
}
